File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/theme/ChanTheme.java
package com.github.k1rakishou.chan.ui.theme;

import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.Typeface;

import com.github.k1rakishou.chan.R;

import static com.github.k1rakishou.chan.utils.AndroidUtils.getRes;

public class ChanTheme extends Theme {

    public ChanTheme(
            String displayName,
            String name,
            int resValue,
            ThemeHelper.PrimaryColor primaryColor
    ) {
        super(displayName, name, resValue, primaryColor);

        resolveSpanColors();
        resolveDrawables();
    }

    public ChanTheme(
            String displayName,
            String name,
            int resValue,
            ThemeHelper.PrimaryColor primaryColor,
            Typeface mainFont,
            Typeface altFont
    ) {
        super(displayName, name, resValue, primaryColor, mainFont, altFont);

        resolveSpanColors();
        resolveDrawables();
    }

    @Override
    public void resolveDrawables() {
        settingsDrawable.tint = Color.BLACK;
        imageDrawable.tint = Color.BLACK;
        sendDrawable.tint = Color.BLACK;
        clearDrawable.tint = Color.BLACK;
        backDrawable.tint = Color.BLACK;
        doneDrawable.tint = Color.BLACK;
        historyDrawable.tint = Color.BLACK;
        helpDrawable.tint = Color.BLACK;
        refreshDrawable.tint = Color.BLACK;
    }

    @SuppressWarnings("ResourceType")
    private void resolveSpanColors() {
        Resources.Theme theme = getRes().newTheme();
        theme.applyStyle(R.style.Chan_DefaultTheme, true);
        theme.applyStyle(resValue, true);

        TypedArray ta = theme.obtainStyledAttributes(new int[]{
                R.attr.post_quote_color,
                R.attr.post_highlight_quote_color,
                R.attr.post_link_color,
                R.attr.post_spoiler_color,
                R.attr.post_inline_quote_color,
                R.attr.post_subject_color,
                R.attr.post_name_color,
                R.attr.post_id_background_light,
                R.attr.post_id_background_dark,
                R.attr.post_capcode_color,
                R.attr.post_details_color,
                R.attr.post_highlighted_color,
                R.attr.post_saved_reply_color,
                R.attr.post_selected_color,
                R.attr.text_color_primary,
                R.attr.text_color_secondary,
                R.attr.text_color_hint,
                R.attr.text_color_reveal_spoiler,
                R.attr.backcolor,
                R.attr.backcolor_secondary,
                R.attr.highlighted_pin_view_holder_color,
                R.attr.pin_posts_not_watching_color,
                R.attr.pin_posts_has_replies_color,
                R.attr.pin_posts_normal_color,
                R.attr.divider_color
        });

        quoteColor = ta.getColor(0, 0);
        highlightQuoteColor = ta.getColor(1, 0);
        linkColor = ta.getColor(2, 0);
        spoilerColor = ta.getColor(3, 0);
        inlineQuoteColor = ta.getColor(4, 0);
        subjectColor = ta.getColor(5, 0);
        nameColor = ta.getColor(6, 0);
        idBackgroundLight = ta.getColor(7, 0);
        idBackgroundDark = ta.getColor(8, 0);
        capcodeColor = ta.getColor(9, 0);
        detailsColor = ta.getColor(10, 0);
        highlightedColor = ta.getColor(11, 0);
        savedReplyColor = ta.getColor(12, 0);
        selectedColor = ta.getColor(13, 0);
        textPrimary = ta.getColor(14, 0);
        textSecondary = ta.getColor(15, 0);
        textHint = ta.getColor(16, 0);
        textColorRevealSpoiler = ta.getColor(17, 0);
        backColor = ta.getColor(18, 0);
        backColorSecondary = ta.getColor(19, 0);
        highlightedPinViewHolderColor = ta.getColor(20, 0);
        pinPostsNotWatchingColor = ta.getColor(21, 0);
        pinPostsHasRepliesColor = ta.getColor(22, 0);
        pinPostsNormalColor = ta.getColor(23, 0);
        dividerColor = ta.getColor(24, 0);

        ta.recycle();
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/theme/ChanTheme.kt
package com.github.k1rakishou.chan.ui.theme

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Typeface
import com.github.k1rakishou.chan.utils.AndroidUtils
import com.github.k1rakishou.model.data.theme.ChanThemeColorId

@SuppressLint("ResourceType")
abstract class ChanTheme {
  abstract val context: Context
  abstract val version: Int
  abstract val name: String
  abstract val isLightTheme: Boolean
  abstract val lightStatusBar: Boolean
  abstract val lightNavBar: Boolean
  abstract val accentColor: Int
  abstract val primaryColor: Int
  abstract val backColor: Int
  abstract val backColorSecondary: Int
  abstract val errorColor: Int
  abstract val textColorPrimary: Int
  abstract val textColorSecondary: Int
  abstract val textColorHint: Int
  abstract val postHighlightedColor: Int
  abstract val postSavedReplyColor: Int
  abstract val postSubjectColor: Int
  abstract val postDetailsColor: Int
  abstract val postNameColor: Int
  abstract val postInlineQuoteColor: Int
  abstract val postQuoteColor: Int
  abstract val postHighlightQuoteColor: Int
  abstract val postLinkColor: Int
  abstract val postSpoilerColor: Int
  abstract val postSpoilerRevealTextColor: Int
  abstract val postUnseenLabelColor: Int
  abstract val dividerColor: Int
  abstract val bookmarkCounterNotWatchingColor: Int
  abstract val bookmarkCounterHasRepliesColor: Int
  abstract val bookmarkCounterNormalColor: Int

  open val mainFont: Typeface = ROBOTO_MEDIUM
  open val altFont: Typeface = ROBOTO_CONDENSED

  val defaultColors by lazy { loadDefaultColors() }

  private fun loadDefaultColors(): DefaultColors {
    val controlNormalColor = if (isLightTheme) {
      CONTROL_LIGHT_COLOR
    } else {
      CONTROL_DARK_COLOR
    }

    val disabledControlAlpha = (255f * .4f).toInt()

    return DefaultColors(disabledControlAlpha, controlNormalColor)
  }

  fun <T : ChanTheme> copy(
    version: Int,
    name: String,
    isLightTheme: Boolean,
    lightStatusBar: Boolean,
    lightNavBar: Boolean,
    accentColor: Int,
    primaryColor: Int,
    backColor: Int,
    backColorSecondary: Int,
    errorColor: Int,
    textPrimaryColor: Int,
    textSecondaryColor: Int,
    textColorHint: Int,
    postHighlightedColor: Int,
    postSavedReplyColor: Int,
    postSubjectColor: Int,
    postDetailsColor: Int,
    postNameColor: Int,
    postInlineQuoteColor: Int,
    postQuoteColor: Int,
    postHighlightQuoteColor: Int,
    postLinkColor: Int,
    postSpoilerColor: Int,
    postSpoilerRevealTextColor: Int,
    postUnseenLabelColor: Int,
    dividerColor: Int,
    bookmarkCounterNotWatchingColor: Int,
    bookmarkCounterHasRepliesColor: Int,
    bookmarkCounterNormalColor: Int,
  ): T {
    return DefaultDarkChanTheme(
      context = context,
      version = version,
      name = name,
      isLightTheme = isLightTheme,
      lightStatusBar = lightStatusBar,
      lightNavBar = lightNavBar,
      accentColor = accentColor,
      primaryColor = primaryColor,
      backColor = backColor,
      backColorSecondary = backColorSecondary,
      errorColor = errorColor,
      textColorPrimary = textPrimaryColor,
      textColorSecondary = textSecondaryColor,
      textColorHint = textColorHint,
      postHighlightedColor = postHighlightedColor,
      postSavedReplyColor = postSavedReplyColor,
      postSubjectColor = postSubjectColor,
      postDetailsColor = postDetailsColor,
      postNameColor = postNameColor,
      postInlineQuoteColor = postInlineQuoteColor,
      postQuoteColor = postQuoteColor,
      postHighlightQuoteColor = postHighlightQuoteColor,
      postLinkColor = postLinkColor,
      postSpoilerColor = postSpoilerColor,
      postSpoilerRevealTextColor = postSpoilerRevealTextColor,
      postUnseenLabelColor = postUnseenLabelColor,
      dividerColor = dividerColor,
      bookmarkCounterNotWatchingColor = bookmarkCounterNotWatchingColor,
      bookmarkCounterHasRepliesColor = bookmarkCounterHasRepliesColor,
      bookmarkCounterNormalColor = bookmarkCounterNormalColor,
    ) as T
  }

  fun getDisabledTextColor(color: Int): Int {
    return if (isLightTheme) {
      AndroidUtils.manipulateColor(color, 1.3f)
    } else {
      AndroidUtils.manipulateColor(color, .7f)
    }
  }

  fun getControlDisabledColor(color: Int): Int {
    return ColorStateList.valueOf(color)
      .withAlpha(defaultColors.disabledControlAlpha)
      .defaultColor
  }

  fun getColorByColorId(chanThemeColorId: ChanThemeColorId): Int {
    return when (chanThemeColorId) {
      ChanThemeColorId.PostSubjectColor -> postSubjectColor
      ChanThemeColorId.PostNameColor -> postNameColor
      ChanThemeColorId.AccentColor -> accentColor
      ChanThemeColorId.PostInlineQuoteColor -> postInlineQuoteColor
      ChanThemeColorId.PostQuoteColor -> postQuoteColor
      ChanThemeColorId.BackColorSecondary -> backColorSecondary
      ChanThemeColorId.PostLinkColor -> postLinkColor
    }
  }

  data class DefaultColors(
    val disabledControlAlpha: Int,
    val controlNormalColor: Int
  ) {

    val disabledControlAlphaFloat: Float
      get() = disabledControlAlpha.toFloat() / MAX_ALPHA_FLOAT

  }

  companion object {
    const val CURRENT_THEME_SCHEMA_VERSION = 1

    private val ROBOTO_MEDIUM = Typeface.create("sans-serif-medium", Typeface.NORMAL)
    private val ROBOTO_CONDENSED = Typeface.create("sans-serif-condensed", Typeface.NORMAL)

    private val CONTROL_LIGHT_COLOR = Color.parseColor("#FFAAAAAA")
    private val CONTROL_DARK_COLOR = Color.parseColor("#FFCCCCCC")

    private const val MAX_ALPHA_FLOAT = 255f
  }
}